util/include/channels/mpsc_bounded_channel.h

```diff
@@ -0,0 +1,238 @@
+    num_senders_++;
+  }
+
+  void removeSender() {
+    bool num_senders = 0;
+    {
+      const std::lock_guard<std::mutex> lock(mutex_);
+      num_senders_--;
+      num_senders = num_senders_;
+    }
+    if (num_senders == 0) {
+      // Signal the receiver to throw a NoSendersError.
+      cv_.notify_one();
+    }
+  }
+
+  void invalidateReceiver() {
+    const std::lock_guard<std::mutex> lock(mutex_);
+    valid_receiver_ = false;
+  }
+
+  size_t capacity_;
+  std::queue<Msg> buf_;
+  mutable std::mutex mutex_;
+  std::condition_variable cv_;
+  size_t num_senders_ = 0;
+  bool valid_receiver_ = true;
+};
+
+}  // namespace impl
+
+template <typename Msg>
+class BoundedMpscSender {
+ public:
+  // Try to send a message over a bounded channel.
+  //
+  // Return std::nullopt on sucess.
+  // Return the Msg if the buffer is full
+  //
+  // Throws NoReceiversError if the receiver has been destructed.
+  //
+  // It's important to note that there is a race condition, whereby the receiver can be destructed
+  // immediately after a successful send. There is no absolute guarantee of delivery. The
+  // purpose of returning an error is to prevent endlessly retrying and never noticing that the
+  // channel is effectively destroyed.
+  std::optional<Msg> send(Msg&& msg) {
+    {
+      std::lock_guard<std::mutex> lock(queue_->mutex_);
+      if (!queue_->valid_receiver_) {
+        throw NoReceiversError();
+      }
+      if (queue_->buf_.size() == queue_->capacity_) {
+        return msg;
+      }
+      queue_->buf_.push(std::move(msg));
+    }
+    queue_->cv_.notify_one();
+
+    return std::nullopt;
+  }
+
+  std::optional<size_t> size() { return queue_->size(); }
+  std::optional<size_t> capacity() { return queue_->capacity(); }
+
+  ~BoundedMpscSender() {
+    if (!queue_) return;
+    queue_->removeSender();
+  }
+
+  BoundedMpscSender(const BoundedMpscSender& other) : BoundedMpscSender(other.queue_) {}
+  BoundedMpscSender& operator=(const BoundedMpscSender& other) {
+    *this = BoundedMpscSender(other);
+    return *this;
+  }
+  BoundedMpscSender(BoundedMpscSender&&) noexcept = default;
+  BoundedMpscSender& operator=(BoundedMpscSender&&) noexcept = default;
+
+ private:
+  BoundedMpscSender(const std::shared_ptr<impl::BoundedMpscQueue<Msg>>& queue) : queue_(queue) { queue_->addSender(); }
+
+  std::shared_ptr<impl::BoundedMpscQueue<Msg>> queue_;
+
+  template <typename T>
+  friend std::pair<BoundedMpscSender<T>, BoundedMpscReceiver<T>> makeBoundedMpscChannel(size_t capacity);
+};  // namespace concord::channel
+
+template <typename Msg>
+class BoundedMpscReceiver {
+ public:
+  // Receive a message over a channel.
+  // Block waiting indefinitely.
+  //
+  // Throws NoSendersError if all senders have been destructed and there are no more messages left
+  // in the communication buffer.
+  Msg recv() {
+    std::unique_lock<std::mutex> lock(queue_->mutex_);
+    if (!queue_->buf_.empty()) {
```

Comment:
Looks like these aren't needed since the predicate form is used and they are coverd in waitPred(). This was a misunderstanding on my part.

Reply:
Fixed.